src/hotspot/share/gc/shenandoah/shenandoahFreeSet.hpp

```diff
@@ -28,69 +29,493 @@
 #include "gc/shenandoah/shenandoahHeapRegionSet.hpp"
 #include "gc/shenandoah/shenandoahHeap.hpp"
 
+// The API and internal implementation of ShenandoahSimpleBitMap and ShenandoahRegionPartitions use ssize_t to
+// represent index, even though index is "inherently" unsigned.  There are two reasons for this choice:
+//  1. We use -1 as a sentinel value to represent empty partitions.  This same value may be used to represent
+//     failure to find a previous set bit or previous range of set bits.
+//  2. Certain loops are written most naturally if the iterator, which may hold the sentinel -1 value, can be
+//     declared as signed and the terminating condition can be < 0.
+
+
+
+// ShenandoahSimpleBitMap resembles CHeapBitMap but adds missing support for find_next_consecutive_bits() and
```

Comment:
Ah I see you have considered using bitMap.* ... but why not 'simply' add the required methods there, instead?

Reply:
Was trying to avoid impacting beyond the boundaries of Shenandoah.  I can tackle this if we think it preferable.  But also trying to work toward "timely integration".

Reply:
Like Roman, I'd prefer extending existing code and/or making it adaptable for our purposes to duplicating functionality. Let us read through the differences and see if this should be done now, or as you state allow this duplication for now in the interests of speed and expediency.

Reply:
If consensus is we should not re-invent the wheel, I'll make an effort to make the additional changes required to use BitMap instead of ShenandoahSimpleBitMap.

Reply:
Thank you Kelvin:

> Here's a rough outline of the differences between ShenandoahSimpleBitMap and src/hotspot/share/utilities/bitMap.*
> 
> 1. ShenandoahSimpleBitMap finds next bit in the forward and backward direction.  BitMap only searches in forward direction.  We use backward searches when we prefer to pack certain types of objects into the high end of the heap.

BitMap could be extended to optionally search backwards as well?

> 2. ShenandoahSimpleBitMap also searches for clusters of bits, to facilitate efficient satisfaction of humongous allocation requests.  BitMap only searches for a single bit.  We could add this capability to BitMap.

I agree.

> 3. If we were to generalize BitMap to handle reverse searches, this may have far-reaching impact on the API.   The type of results and idx_t (typedef size_t).  When we do backward searches, ShenandoahSimpleBitMap treats -1 as the sentinel NOT_FOUND value.

I think that's easy to fix. Backwards searches from index x (exclusive), can use `find_last_set_bit(idx_t 0, idx_t x)`. If you get back `x` that indicates NOT_FOUND:

```
  // Return the index of the last set (or clear) bit in the range [beg, end),
  // or end if none found.
  // precondition: beg and end form a valid range for the bitmap.
  idx_t find_last_set_bit(idx_t beg, idx_t end) const;
  idx_t find_last_clear_bit(idx_t beg, idx_t end) const;
```

You can use a wrapper to return `-1` if you want for your specific use case.

Would that work?

PS: I have not looked at `cluster of bits`, but I am guessing it's not a bit but a specific pattern/sequence you are looking for. I'll take a look at the code of `ShenandoahSimpleBitMap` to understand.

Reply:
I'd say, let's keep the duplicate ShSimpleBitMap for now, but make a note (bug-entry?) to attempt a merge later on. We can help that by trying to keep the style, types, algorithmic and API approaches similar.